engine/src/main/java/org/hibernate/search/engine/metadata/impl/AnnotationMetadataProvider.java

```diff
@@ -1751,4 +1779,22 @@ boolean stateInspectionOptimizationsEnabled(TypeMetadata.Builder typeMetadataBui
 		metadataProvidingFieldBridge.configureFieldMetadata( fieldName, builder );
 		return builder.getSortableFields();
 	}
+
+	private boolean hasExplicitDocumentId(XClass type) {
```

Comment:
I am not particularly happy with that additional pass, but I cannot think of a way for avoiding it without re-writing large parts here. We'd have to delay the validation against id default field redefinition from retrieval of the annotation to a later point where we already have visited all properties. But even then we'd have to do another pass of the previously built model, so likely not much would be gained.

Also I am avoiding the repeated class loading of the id types by pulling that out, so that should compensate a bit ;)